pennylane_forest/converter.py

```diff
@@ -0,0 +1,626 @@
+        """Instantiate the parametrized gate.
+        
+        Args:
+            variable_map (Dict[str, object]): The map that assigns values to variable names
+            qubit_to_wire_map ([type]): The map that assigns wires to qubits
+        
+        Returns:
+            qml.Operation: The initiated instance of the parametrized PennyLane gate
+        """
+        resolved_params = _resolve_params(self.pyquil_params, variable_map)
+        resolved_wires = _qubits_to_wires(self.pyquil_qubits, qubit_to_wire_map)
+
+        pl_gate_instance = self.pl_gate(*resolved_params, wires=resolved_wires)
+
+        if self.is_inverted:
+            pl_gate_instance.inv()
+
+        return pl_gate_instance
+
+
+class ParametrizedQubitUnitary:
+    """Represents a QubitUnitary instance already parametrized with a matrix.
+    
+    Args:
+        matrix (np.array): The unitary gate matrix
+    """
+
+    def __init__(self, matrix):
+        self.matrix = matrix
+
+    def __call__(self, wires):
+        """Instantiate the QubitUnitary on the given wires.
+        
+        Args:
+            wires (List[int]): The wires the QubitUnitary acts on
+        
+        Returns:
+            qml.QubitUnitary: The instantiate QubitUnitary instance
+        """
+        return qml.QubitUnitary(self.matrix, wires=wires)
+
+
+class ProgramLoader:
+    """Loads the given pyquil Program as a PennyLane template.
+
+    The pyquil Program is parsed once at instantiation and can be 
+    applied as often as desired.
+        
+    Args:
+        program (pyquil.quil.Program): The pyquil Program instance that should be loaded
+    """
+
+    _matrix_dictionary = pyquil.gate_matrices.QUANTUM_GATES
+
+    def __init__(self, program):
+        self.program = program
+        self.qubits = program.get_qubits()
+
+        self._load_defined_gate_names()
+        self._load_declarations()
+        self._load_measurements()
+        self._load_template()
+
+    def _load_defined_gate_names(self):
+        """Extract the names of all defined gates of the pyquil Program."""
+        self._defined_gate_names = []
+
+        for defgate in self.program.defined_gates:
+            self._defined_gate_names.append(defgate.name)
+
+            if isinstance(defgate, pyquil.quil.DefPermutationGate):
+                matrix = np.eye(defgate.permutation.shape[0])
+                matrix = matrix[:, defgate.permutation]
+            elif isinstance(defgate, pyquil.quil.DefGate):
+                matrix = defgate.matrix
+
+            self._matrix_dictionary[defgate.name] = matrix
+
+    def _load_declarations(self):
+        """Extract the declarations of the pyquil Program."""
+        self._declarations = [
+            instruction for instruction in self.program.instructions if _is_declaration(instruction)
+        ]
+
+    def _load_measurements(self):
+        """Extract the measurements of the pyquil Program."""
+        self._measurements = [
```

Comment:
Could this be simply just a local object (`_measurements`) instead of an instance attribute? There seem to be no more occurrences for `self._measurements` other than this one

Reply:
Good catch!